include/pybind11/pytypes.h

```diff
@@ -1331,35 +1332,109 @@ class buffer : public object {
+        m_ptr = (info.view()->obj) ?
+            PyMemoryView_FromObject(info.view()->obj) :
+            PyMemoryView_FromBuffer(info.view());
         if (!m_ptr)
             pybind11_fail("Unable to create memoryview from buffer descriptor");
     }
 
-    PYBIND11_OBJECT_CVT(memoryview, object, PyMemoryView_Check, PyMemoryView_FromObject)
+    /** \rst
+        Creates ``memoryview`` from static buffer.
+
+        The caller is responsible for managing the lifetime of ``ptr`` and
+        ``format``. This method is meant for providing a ``memoryview`` for
+        C/C++ buffer not managed by Python.
+
+        :param ptr: Pointer to the buffer.
+        :param itemsize: Byte size of an element.
+        :param format: Pointer to the null-terminated format string. For
+            homogeneous Buffers, this should be set to
+            format_descriptor<T>::value.
+        :param ndim: Number of dimensions.
+        :param shape_in: Shape of the tensor (1 entry per dimension).
+        :param strides_in: Number of bytes between adjacent entries (for each
+            per dimension).
+        :param readonly: Flag to indicate if the underlying storage may be
+            written to.
+     \endrst */
+    static memoryview frombuffer(
+        void *ptr, ssize_t itemsize, const char* format, ssize_t ndim,
+        detail::any_container<ssize_t> shape_in,
+        detail::any_container<ssize_t> strides_in, bool readonly=false);
+
+    template<typename T>
+    static memoryview frombuffer(
+        T *ptr, detail::any_container<ssize_t> shape_in,
+        detail::any_container<ssize_t> strides_in, bool readonly=false) {
+        return memoryview::frombuffer(
+            reinterpret_cast<void*>(ptr), sizeof(T),
+            format_descriptor<T>::value, static_cast<ssize_t>(shape_in->size()),
+            shape_in, strides_in, readonly);
+    }
+
+#if PY_MAJOR_VERSION >= 3
+    /** \rst
+        Creates ``memoryview`` from static memory.
+
+        The caller is responsible for managing the lifetime of ``mem``. This
+        constructor is meant for providing a ``memoryview`` for C/C++ buffer not
+        managed by Python.
+     \endrst */
+    static memoryview frommemory(char *mem, ssize_t size, bool writable = false) {
+        PyObject* ptr = PyMemoryView_FromMemory(
+            mem, size, (writable) ? PyBUF_WRITE : PyBUF_READ);
+        if (!ptr)
+            pybind11_fail("Could not allocate memoryview object!");
+        return memoryview(object(ptr, stolen_t{}));
+    }
+#endif
 };
+
+inline memoryview memoryview::frombuffer(
+    void *ptr, ssize_t itemsize, const char* format, ssize_t ndim,
+    detail::any_container<ssize_t> shape_in,
+    detail::any_container<ssize_t> strides_in, bool readonly) {
+    std::vector<Py_ssize_t> shape(std::move(shape_in));
+    std::vector<Py_ssize_t> strides(std::move(strides_in));
+    if (ndim != static_cast<ssize_t>(shape.size()) ||
```

Comment:
Could this be moved to the top, with `shape_in->size()` and `shape_out->size()`?
Could you just use e.g. `shape_in->data` below instead of copying/moving into your own `std::vector`?
`any_container` contains a `std::vector` already and has `const std::vector<T> *operator->() const`.

Reply:
Ok